src/gp_engine.hpp

```diff
@@ -0,0 +1,132 @@
+
+class GPEngine {
+ public:
+  GPEngine(){};
+  GPEngine(SitePattern site_pattern, size_t pcss_count);
+
+  // These operators mean that we can invoke this class on each of the operations.
+  void operator()(const GPOperations::Zero& op);
+  void operator()(const GPOperations::SetToStationaryDistribution& op);
+  void operator()(const GPOperations::WeightedSumAccumulate& op);
+  void operator()(const GPOperations::Multiply& op);
+  void operator()(const GPOperations::Likelihood& op);
+  void operator()(const GPOperations::EvolveRootward& op);
+  void operator()(const GPOperations::EvolveLeafward& op);
+  void operator()(const GPOperations::OptimizeRootward& op);
+  void operator()(const GPOperations::OptimizeLeafward& op);
+  void operator()(const GPOperations::UpdateSBNProbabilities& op);
+
+  void ProcessOperations(GPOperationVector operations);
+
+  void SetTransitionMatrixToHaveBranchLength(double branch_length);
+  void SetTransitionAndDerivativeMatricesToHaveBranchLength(double branch_length);
+  void SetTransitionMatrixToHaveBranchLengthAndTranspose(double branch_length);
+  const Eigen::Matrix4d& GetTransitionMatrix() { return transition_matrix_; };
+  void PrintPLV(size_t plv_idx);
+
+  void SetBranchLengths(EigenVectorXd branch_lengths) {
+    branch_lengths_ = branch_lengths;
+  };
+  EigenVectorXd GetBranchLengths() const { return branch_lengths_; };
+  EigenVectorXd GetLogLikelihoods() const { return log_likelihoods_; };
+
+  DoublePair LogLikelihoodAndDerivative(const GPOperations::OptimizeRootward& op);
+
+ private:
+  double min_branch_length_ = 1e-6;
+  double max_branch_length_ = 3.;
+  int significant_digits_for_optimization_ = 6;
+  double relative_tolerance_for_optimization_ = 1e-2;
+  double step_size_for_optimization_ = 5e-4;
+  size_t max_iter_for_optimization_ = 100;
+
+  SitePattern site_pattern_;
+  std::vector<NucleotidePLV> plvs_;
+  EigenVectorXd branch_lengths_;
+  EigenVectorXd log_likelihoods_;
+  EigenVectorXd q_;
+
+  // Internal "temporaries" useful for likelihood and derivative calculation.
+  EigenVectorXd per_pattern_log_likelihoods_;
+  EigenVectorXd per_pattern_likelihoods_;
+  EigenVectorXd per_pattern_likelihood_derivatives_;
+  EigenVectorXd per_pattern_likelihood_derivative_ratios_;
+
+  // When we change from JC69Model, check that we are actually doing transpose in
+  // leafward calculations.
+  JC69Model substitution_model_;
+  Eigen::Matrix4d eigenmatrix_ = substitution_model_.GetEigenvectors().reshaped(4, 4);
+  Eigen::Matrix4d inverse_eigenmatrix_ =
+      substitution_model_.GetInverseEigenvectors().reshaped(4, 4);
+  Eigen::Vector4d eigenvalues_ = substitution_model_.GetEigenvalues();
+  Eigen::DiagonalMatrix<double, 4> diagonal_matrix_;
+  Eigen::Matrix4d transition_matrix_;
+  Eigen::Matrix4d derivative_matrix_;
+  Eigen::Vector4d stationary_distribution_ = substitution_model_.GetFrequencies();
+  EigenVectorXd site_pattern_weights_;
```

Comment:
I guess I'm not familiar with the terminology "site pattern".

Reply:
It's a version of a sequence alignment that compresses columns with the same likelihood https://github.com/phylovi/libsbn/blob/master/src/site_pattern.hpp